src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -129,6 +246,172 @@ private void ParseMarkupText(in SyntaxListBuilder<RazorSyntaxNode> builder)
             AcceptAndMoveNext();
         }
 
+        private void ParseMarkupInCodeBlock(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            do
+            {
+                switch (GetParserState(ParseMode.MarkupInCodeBlock))
+                {
+                    case ParserState.EOF:
+                        CompleteMarkupInCodeBlock(builder);
+                        return;
+                    case ParserState.Tag:
+                        ParseMarkupElement(builder, ParseMode.MarkupInCodeBlock);
+                        break;
+                    case ParserState.SpecialTag:
+                    case ParserState.XmlPI:
+                    case ParserState.MarkupComment:
+                    case ParserState.CData:
+                        ParseMarkupNode(builder, ParseMode.MarkupInCodeBlock);
+                        SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.None;
+                        builder.Add(OutputAsMarkupLiteral());
+                        break;
+                    default:
+                        ParseMarkupNode(builder, ParseMode.Text);
+                        break;
+                }
+            } while (_tagTracker.Count > 0);
+
+            CompleteMarkupInCodeBlock(builder);
+        }
+
+        private void CompleteMarkupInCodeBlock(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            // Output anything we've accepted so far.
+            builder.Add(OutputAsMarkupLiteral());
+
+            var completeOuterTag = true;
+            while (_tagTracker.Count > 0)
+            {
+                var tracker = _tagTracker.Pop();
+                var element = SyntaxFactory.MarkupElement(tracker.StartTag, builder.Consume(), endTag: null);
+                builder.AddRange(tracker.PreviousNodes);
+                builder.Add(element);
+
+                if (_tagTracker.Count == 0)
+                {
+                    completeOuterTag = tracker.IsComplete;
+                    if (completeOuterTag)
+                    {
+                        // We're at the outermost start tag. Add an error.
+                        // We don't want to add this error if the tag is unfinished. A different error would have already been added.
+                        Context.ErrorSink.OnError(
+                            RazorDiagnosticFactory.CreateParsing_MissingEndTag(
+                                new SourceSpan(
+                                    SourceLocationTracker.Advance(tracker.TagLocation, "<"),
+                                    tracker.TagName.Length),
+                                tracker.TagName));
+                    }
+                }
+            }
+
+            if (!Context.DesignTimeMode)
```

Comment:
I'm not totally certain that I know what's going on this in part.

Reply:
I'll add a comment.
E.g,
```HTML
@{
    <div>Foo</div>|     |
    <text>Bar</text>|      |
}
```
Basically, this part decides what happens to the whitespace after we are done parsing a block of markup.
In the first case, the whitespace should be accepted as markup. In the second case, it should be code.
Unrelated to this, we also need to figure out what accepted characters to assign to this based on whether a tag is well-formed or not.